### PR TITLE
feat: support bff operators

### PR DESCRIPTION
**Pull request by yimingjfe**
# PR Details

提供一系列的 BFF 操作符，拓展现有 BFF 语法，提升 BFF 的易用度和功能，该 feature 无 break change，暂时不更新官网文档，先做内部验证。

## Description

基本使用示例如下，服务端通过 `Api` 函数定义接口与类型，接受到的数据会自动地进行数据校验。

```typescript title="api/user.ts"
import { Api } from '@modern-js/server/runtime'
import { z } from 'zod'

const UserSchema = z.object({
  name: z.string().min(2).max(10),
  email: z.string().email()
});

const DataSchema = z.object({
  phone: z.string()
})

const getUser = Api(
  Get("/user"),
  Query(UserSchema),
  Data(DataSchema),
  async ({ query, data }) => {
    return {
      name: query.name,
      phone: data.phone
    }
  }
)
```

前端根据服务端定义的类型，做一体化调用：
```typescript title="src/App.tsx"
import { getUser } from '@api/user'

get({
    query: {
        name: 'modern-js',
        email: 'modern-js@github.com'
    },
    data: {
      phone: 12345
    },
})
```

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/server/bff-core/src/operators/http.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+import { z } from 'zod';
```
Comment (zllkjc) [inline packages/server/bff-core/src/operators/http.ts:1]: 这里不能这样吧，如果是 peer 的话

File: packages/server/bff-core/src/api.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      helper.result = res;
+      return next();
+    });
+
+    await compose(stack)(executeHelper);
+
+    return executeHelper.result;
+  }
+
+  (runner as Runner)[HANDLER_WITH_META] = true;
```
Comment (zllkjc) [inline packages/server/bff-core/src/api.ts:77]: 这边类型变来变去的

File: packages/cli/plugin-bff/package.json [hunk 3]
```diff
@@ -61,13 +61,13 @@
...
   },
   "dependencies": {
-    "@babel/core": "^7.18.0",
-    "@babel/runtime": "^7.18.0",
-    "@modern-js/babel-compiler": "workspace:^1.2.6",
-    "@modern-js/bff-utils": "workspace:^1.2.9",
-    "@modern-js/create-request": "workspace:^1.2.11",
-    "@modern-js/server-utils": "workspace:^1.2.11",
-    "@modern-js/utils": "workspace:^1.7.7"
+    "@babel/core": "^7.17.0",
```
Comment (chenjiahan) [inline packages/cli/plugin-bff/package.json:64]: 这里版本降级的原因是啥

File: packages/cli/plugin-bff/package.json [hunk 4]
```diff
@@ -11,7 +11,7 @@
     "modern",
     "modern.js"
   ],
-  "version": "1.5.2",
+  "version": "1.5.3-beta.1",
```
Comment (zllkjc) [inline packages/cli/plugin-bff/package.json:14]: 这里要不改回去

File: packages/server/bff-core/package.json [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  "dependencies": {
+    "@babel/runtime": "^7.15.4",
+    "@modern-js/bff-runtime": "workspace:^1.2.3",
+    "@modern-js/utils": "workspace:^1.7.6",
+    "esbuild": "^0.14.38",
+    "esbuild-register": "^3.3.3",
+    "http-errors": "^2.0.0",
+    "koa-compose": "^4.1.0",
+    "reflect-metadata": "^0.1.13"
```
Comment (chenjiahan) [inline packages/server/bff-core/package.json:41]: 建议做一下依赖预打包

**Comment by chenjiahan** [comment]
能否通过 QoS 验证本次改动对 QPS 的影响

File: packages/server/bff-core/src/client/result.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+export type Err<T = any> = {
+  kind: 'Err';
+  value: T;
+  isErr: true;
+  isOk: false;
+};
+
+export type Ok<T = any> = {
```
Comment (chenjiahan) [inline packages/server/bff-core/src/client/result.ts:8]: `any` -> `unknown` ?

File: packages/server/bff-core/src/api.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    const executeHelper: ExecuteHelper = {
+      result: null,
+      get inputs() {
+        return inputs;
+      },
+    };
+
+    const stack = [...validateHandlers];
+
+    stack.push(async (helper: ExecuteHelper, next: any) => {
```
Comment (chenjiahan) [inline packages/server/bff-core/src/api.ts:66]: no any

File: packages/cli/plugin-bff/package.json (hunk 3 above)
Comment (yimingjfe) [reply packages/cli/plugin-bff/package.json:64]: 我看一下

File: packages/review/testing-plugin-bff/package.json [hunk 8]
```diff
@@ -36,24 +36,26 @@
     "test": "wireit"
   },
   "dependencies": {
-    "@babel/runtime": "^7.18.0",
-    "@modern-js/server": "workspace:^1.4.20",
-    "@modern-js/utils": "workspace:^1.7.7",
+    "@babel/runtime": "^7",
+    "@modern-js/bff-core": "workspace:^1.0.0",
```
Comment (zllkjc) [inline packages/review/testing-plugin-bff/package.json:40]: 依赖降级了

File: packages/review/testing-plugin-bff/package.json [hunk 9]
```diff
@@ -36,24 +36,26 @@
...
+    "@modern-js/server": "workspace:^1.4.16",
+    "@modern-js/utils": "workspace:^1.7.6",
```
Comment (zllkjc) [inline packages/review/testing-plugin-bff/package.json:42]: 依赖降了

Comment (zllkjc) [reply packages/review/testing-plugin-bff/package.json:42]: 这个应该不影响

**Review by zllkjc — APPROVED** [review]

**Pull request merged**